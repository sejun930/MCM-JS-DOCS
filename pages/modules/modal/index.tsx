import MyModal from "src/main/mainComponents/modules/modal";

export default function _Modal() {
  return <MyModal />;
}
