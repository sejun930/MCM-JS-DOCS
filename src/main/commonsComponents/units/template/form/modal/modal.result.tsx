import styled from "@emotion/styled";
import { ReactNode } from "react";

export default function ModalResultForm({
  children, // children 렌더
}: {
  children: ReactNode;
}) {
  return <Wrapper>{children}</Wrapper>;
}

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;
  width: 100%;
  height: 100%;
`;

export const ButtonWrapper = styled.div`
  display: flex;

  .cmm-button-unit {
    padding: 10px 16px;
    border-radius: 10px;
    font-size: 16px;
    color: #666666;
  }
`;
