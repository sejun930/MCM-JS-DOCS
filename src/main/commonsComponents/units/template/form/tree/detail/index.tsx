import {
  Wrapper,
  DetailInfoWrapper,
  DetailTitle,
  DetailTitleInfo,
  DetailRole,
  DetailWrapper,
  ParentsWrapper,
  EmptySelect,
} from "./tree.datail.styles";
import { TreeModuleListTypes } from "src/commons/data/tree/tree.commons.data";

import { _PText, _SpanText, _PTextWithHtml, _Button } from "mcm-js-commons";
import { TreeIProps } from "..";

export default function ModuleTreeDetailPage({
  treeList,
  select,
  selectTree,
}: TreeIProps) {
  // 현재 선택된 정보 가져오기
  const detailInfo = treeList[select];

  let parentsInfo: Partial<TreeModuleListTypes> = {};
  let parentIdx = 0;

  if (treeList.length && detailInfo?.depth) {
    const filterArr = treeList.slice(0, select + 1).filter((el, i) => {
      if (detailInfo?.depth - 1 === el.depth) {
        parentIdx = i;
        return true;
      }
      return false;
    });

    if (filterArr.at(-1)?.class) {
      parentsInfo = filterArr?.at(-1) || {};
    }
  }

  return (
    <Wrapper className="tree-detail-wrapper">
      {!detailInfo ? (
        <EmptySelect className="tree-detail-empty-select">
          <_PText>태그를 클릭해보세요.</_PText>
        </EmptySelect>
      ) : (
        <DetailWrapper className="tree-detail-contents-wrapper">
          <DetailInfoWrapper className="tree-detail-contents-info-wrapper">
            <DetailTitleInfo className="tree-detail-title-wrapper">
              <DetailTitle className="tree-detail-tag-type">
                <_SpanText>Tag Type</_SpanText>
                <code
                  dangerouslySetInnerHTML={{
                    __html: `<<b class='darkBlue'>${detailInfo.tag}</b>>`,
                  }}
                />
              </DetailTitle>
              <DetailTitle className="tree-detail-tag-className">
                <_SpanText>ClassName</_SpanText>
                <code
                  dangerouslySetInnerHTML={{
                    __html: `<b class='lightOrange'>${detailInfo.class}</b>`,
                  }}
                />
              </DetailTitle>
            </DetailTitleInfo>
            <DetailRole className="tree-detail-role-wrapper">
              <_PTextWithHtml
                className="tree-detail-info-role"
                dangerouslySetInnerHTML={detailInfo?.role || "-"}
              />
            </DetailRole>
          </DetailInfoWrapper>
          <ParentsWrapper>
            <_Button
              onClickEvent={() => selectTree(parentIdx)}
              className="tree-detail-button"
            >
              <_PText className="my-parents">
                상위 태그 :
                {(parentsInfo.class && (
                  <code
                    dangerouslySetInnerHTML={{
                      __html: `<<b class="darkBlue">${parentsInfo.tag}</b> <span class='skyblue'>class=</span><span class='lightOrange' style="font-weight : 700">"${parentsInfo.class}"</span> />`,
                    }}
                  />
                )) ||
                  " - "}
              </_PText>
            </_Button>
          </ParentsWrapper>
        </DetailWrapper>
      )}
    </Wrapper>
  );
}
