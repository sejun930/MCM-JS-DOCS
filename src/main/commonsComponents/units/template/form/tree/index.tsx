import styled from "@emotion/styled";
import { breakPoints } from "mcm-js-commons/dist/responsive";

import { useState } from "react";
import { useRecoilState } from "recoil";
import { moduleState } from "src/commons/store";

import { Wrapper } from "../form.commons.styles";

import {
  TreeModuleListTypes,
  treeModuleList,
} from "src/commons/data/tree/tree.commons.data";

import _SubTitleTemplate from "../../title/subTitle";
import ModuleTreeListPage from "./list";
import ModuleTreeDetailPage from "./detail";

// 폴더 구조 예시용 폼
export default function _TreeForm() {
  // 선택한 데이터
  const [select, setSelect] = useState<number>(0);
  const [module] = useRecoilState(moduleState);

  const list = treeModuleList[module] || [];

  // 마우스 올렸을 경우 해당 데이터 선택하기
  const selectTree = (num: number) => {
    setSelect(num === select ? -1 : num);
  };

  return (
    <Wrapper id="tree-form">
      <_SubTitleTemplate
        title="Module Tree"
        className="tree-subTitle"
        remakrs="모듈의 구조 안의 각각의 태그들이 어떤 역할을 하고 있는지 태그를 클릭해보세요."
      />
      <TreeWrapper>
        {list && (
          <>
            <ModuleTreeListPage
              treeList={list}
              selectTree={selectTree}
              select={select}
            />
            <ModuleTreeDetailPage
              treeList={list}
              selectTree={selectTree}
              select={select}
            />
          </>
        )}
      </TreeWrapper>
    </Wrapper>
  );
}

export interface TreeIProps {
  treeList: Array<TreeModuleListTypes>;
  selectTree: (num: number) => void;
  select: number;
}

export const TreeWrapper = styled.div`
  display: flex;
  margin-top: 30px;
  border: solid 1px #777777;
  width: 100%;
  height: 300px;
  border-radius: 10px;
  position: relative;

  @media ${breakPoints.mobileLarge} {
    height: auto;
    flex-direction: column;
  }
`;
