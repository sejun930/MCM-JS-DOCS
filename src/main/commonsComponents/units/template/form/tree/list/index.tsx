import { TreeListWrapper } from "./list.styles";

import _Copy from "src/main/commonsComponents/units/copy";
import { TreeIProps } from "..";

declare const window: typeof globalThis & {
  selectTree: (i: number) => void;
};

export default function ModuleTreeListPage({
  treeList,
  selectTree,
  select,
}: TreeIProps) {
  const getList = (): string => {
    return treeList.reduce((acc, cur, i): string => {
      if (typeof window !== "undefined")
        window.selectTree = function (i: number) {
          selectTree(i);
        };

      let str: string = `<code style="--left : ${
        cur.depth
      }" onmouseover="selectTree(${i})"}
        class="${(select === i && "select-tree") || ""}"
      >`;
      str += `${
        (cur.depth && "<span>âˆŸ</span> ") || ""
      }<<span class='darkBlue'>${
        cur.tag
      } <span class='skyblue'>class=</span><span class='lightOrange'>"${
        cur.class
      }"</span></span>>`;

      str += `</code>`;

      return acc + str;
    }, "");
  };

  return (
    <TreeListWrapper>
      <_Copy type="Code" text={getList()} className="tree-list" copyDisable />
      {/* <TreeListItems dangerouslySetInnerHTML={{ __html: getList() }} /> */}
    </TreeListWrapper>
  );
}
