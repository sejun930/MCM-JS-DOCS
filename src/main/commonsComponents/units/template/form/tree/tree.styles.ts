import styled from "@emotion/styled";

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

export const TreeWrapper = styled.div`
  display: flex;
  margin-top: 30px;
  border: solid 1px #777777;
  width: 100%;
  height: 300px;
  border-radius: 10px;
  position: relative;
`;
