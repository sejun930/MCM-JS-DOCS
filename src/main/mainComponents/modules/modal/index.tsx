import { useState } from "react";

import styled from "@emotion/styled";
import {
  modalExampleList,
  modalExampleInitProps,
} from "./example/modal.example.render.data";
import { modalCodeList } from "./example/modal.example.code.data";

import Template from "src/main/commonsComponents/units/template/main";
import _MainTitleTemplate from "src/main/commonsComponents/units/template/title/mainTitle";
import _HowUseForm from "src/main/commonsComponents/units/template/form/howUse";
import _ExampleForm from "src/main/commonsComponents/units/template/form/example";
import CommonsHooksComponents from "src/main/commonsComponents/hooks/commonsHooks";

export default function MyModal() {
  // 모달을 오픈할 show state (true일 때 모달 오픈)
  const { getAllExampleComponentLength } = CommonsHooksComponents();

  const [isShow, setIsShow] = useState(
    new Array(getAllExampleComponentLength(modalExampleList)).fill(false)
  );

  // 버튼 클릭 시 모달 오픈
  const openModal = (idx: number) => (): void => {
    const temp = [...isShow];
    temp[idx] = true;

    setIsShow(temp);
  };

  // 모달 종료 함수
  const closeModal = (idx: number) => (): void => {
    const temp = [...isShow];
    temp[idx] = false;

    setIsShow(temp);
  };

  // 예시용 컴포넌트로 전달되는 props
  const commonsProps = { isShow, openModal, closeModal };

  return (
    <Template>
      <_MainTitleTemplate />
      <ModulesInfoWrapper>
        <_HowUseForm
          code={modalCodeList.basic}
          exmapleContents="기본 모달 페이지입니다."
        />
        <_ExampleForm
          exampleList={modalExampleList}
          initProps={modalExampleInitProps}
          commonsProps={commonsProps}
        />
      </ModulesInfoWrapper>
    </Template>
  );
}

export const ModulesInfoWrapper = styled.div`
  display: flex;
  flex-direction: column;
  gap: 100px 0px;
`;
