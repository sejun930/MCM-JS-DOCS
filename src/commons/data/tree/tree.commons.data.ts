import { modalTreeList } from "src/main/mainComponents/modules/modal/tree/modal.tree";
import { tooltipTreeList } from "src/main/mainComponents/modules/tooltip/tree/tooltip.tree";
import { sliderTreeList } from "src/main/mainComponents/modules/slider/tree/slider.tree";
import { alertTreeList } from "src/main/mainComponents/modules/alert/tree/alert.tree";

export interface TreeModuleListTypes {
  tag: string; // 태그 타입
  class: string; // 클래스 네임
  depth: number; // 태그의 깊이 (높을 수록 안쪽에 배치 됨)
  role: string; // 해당 태그의 역할 및 부가설명
}

export const treeModuleList: { [key: string]: Array<TreeModuleListTypes> } = {
  Modal: modalTreeList,
  Tooltip: tooltipTreeList,
  Slider: sliderTreeList,
  Alert: alertTreeList,
};
