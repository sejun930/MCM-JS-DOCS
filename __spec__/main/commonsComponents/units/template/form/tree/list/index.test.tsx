import { render } from "@testing-library/react";

import { modalTreeList } from "src/main/mainComponents/modules/modal/tree/modal.tree";
import ModuleTreeListPage from "src/main/commonsComponents/units/template/form/tree/list";

export default describe("Tree List Page", () => {
  // 스냅샷
  test("Snapshot", () => {
    const { container } = render(
      <ModuleTreeListPage
        treeList={modalTreeList}
        select={0}
        selectTree={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });

  // wrapper 및 copy가 있는지 검증
  test("Check wrapper and copy", () => {
    const { container } = render(
      <ModuleTreeListPage
        treeList={modalTreeList}
        select={0}
        selectTree={() => {}}
      />
    );

    const wrapperEle = container.querySelector(".tree-list-wrapper");
    expect(wrapperEle).toBeInTheDocument();

    if (wrapperEle) {
      const copyEle = wrapperEle.querySelector(".tree-list");
      expect(copyEle).toBeInTheDocument();
    }
  });
});
