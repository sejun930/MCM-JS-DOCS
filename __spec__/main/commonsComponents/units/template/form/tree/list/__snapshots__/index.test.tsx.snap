// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree List Page Snapshot 1`] = `
<div>
  <div
    class="tree-list-wrapper css-1hxre6i"
  >
    <div
      class="copy-wrapper tree-list css-di3nu0"
    >
      <div
        class="copy-text css-gjwso2"
      >
        <pre
          class="copy-code-list css-1gnhfxs"
        >
          <code
            class="copy-code"
          >
            <code
              class="select-tree"
              onclick="selectTree(0)"
              style="--left : 0"
              }=""
            >
              &lt;
              <span
                class="darkBlue"
              >
                div 
                <span
                  class="skyblue"
                >
                  class=
                </span>
                <span
                  class="lightOrange"
                >
                  "mcm-modal-wrapper"
                </span>
              </span>
              &gt;
            </code>
            <code
              class=""
              onclick="selectTree(1)"
              style="--left : 1"
              }=""
            >
              <span>
                ∟
              </span>
               &lt;
              <span
                class="darkBlue"
              >
                div 
                <span
                  class="skyblue"
                >
                  class=
                </span>
                <span
                  class="lightOrange"
                >
                  "mcm-modal-items"
                </span>
              </span>
              &gt;
            </code>
            <code
              class=""
              onclick="selectTree(2)"
              style="--left : 2"
              }=""
            >
              <span>
                ∟
              </span>
               &lt;
              <span
                class="darkBlue"
              >
                div 
                <span
                  class="skyblue"
                >
                  class=
                </span>
                <span
                  class="lightOrange"
                >
                  "mcm-modal-close-button-wrapper"
                </span>
              </span>
              &gt;
            </code>
            <code
              class=""
              onclick="selectTree(3)"
              style="--left : 3"
              }=""
            >
              <span>
                ∟
              </span>
               &lt;
              <span
                class="darkBlue"
              >
                button 
                <span
                  class="skyblue"
                >
                  class=
                </span>
                <span
                  class="lightOrange"
                >
                  "mcm-modal-close-ment-button"
                </span>
              </span>
              &gt;
            </code>
            <code
              class=""
              onclick="selectTree(4)"
              style="--left : 4"
              }=""
            >
              <span>
                ∟
              </span>
               &lt;
              <span
                class="darkBlue"
              >
                span 
                <span
                  class="skyblue"
                >
                  class=
                </span>
                <span
                  class="lightOrange"
                >
                  "mcm-modal-close-ment"
                </span>
              </span>
              &gt;
            </code>
            <code
              class=""
              onclick="selectTree(5)"
              style="--left : 3"
              }=""
            >
              <span>
                ∟
              </span>
               &lt;
              <span
                class="darkBlue"
              >
                button 
                <span
                  class="skyblue"
                >
                  class=
                </span>
                <span
                  class="lightOrange"
                >
                  "mcm-modal-close-button"
                </span>
              </span>
              &gt;
            </code>
            <code
              class=""
              onclick="selectTree(6)"
              style="--left : 2"
              }=""
            >
              <span>
                ∟
              </span>
               &lt;
              <span
                class="darkBlue"
              >
                div 
                <span
                  class="skyblue"
                >
                  class=
                </span>
                <span
                  class="lightOrange"
                >
                  "mcm-modal-layout"
                </span>
              </span>
              &gt;
            </code>
            <code
              class=""
              onclick="selectTree(7)"
              style="--left : 3"
              }=""
            >
              <span>
                ∟
              </span>
               &lt;
              <span
                class="darkBlue"
              >
                div 
                <span
                  class="skyblue"
                >
                  class=
                </span>
                <span
                  class="lightOrange"
                >
                  "mcm-modal-content"
                </span>
              </span>
              &gt;
            </code>
          </code>
        </pre>
      </div>
    </div>
  </div>
</div>
`;
